Implement ATen Distributions.cu + Poisson

rachtsingh opened the pull request:
I got a bit into the weeds and the scope expanded, but I think this is the right way to do it. This PR (will) implement a pointwise Poisson sampling method for CPU/CUDA, and should be followed up quickly with ports of the Gamma / Dirichlet samplers, and other things we might want to use from `distributions.c`.

So far:
- [x] Renames `Generator` inside `THCRandomTensor.h` to be `THCGenerator`, and fix CUDAGenerator instantiation
- [x] Implement Poisson distribution + `torch.poisson` (Variable only for now, but we can wrap it). 
- [x] Write tests for the distribution.
- [x] Implement a Philox-based sampler for rejection sampling.

rachtsingh commented:
Ah, not quite ready for review yet, by the way - I still need to debug this CUDA issue and write tests.

fritzo commented on aten/src/THC/THCTensorRandom.cu:
```diff
@@ -97,34 +175,48 @@ __device__ inline half half_uniform_scale_and_shift(float x, double a, double b)
-      result[i] = y;                                                           \
-    }                                                                          \
-  }                                                                            \
+#define GENERATE_KERNEL1(NAME, T, ARG1, CURAND_T, CURAND_FUNC, TRANSFORM)           \
```
nit: Consider reverting unnecessary whitespace changes to ease review by PyTorch folks.

rachtsingh replied:
Thanks! I usually squash before review but I'll be more on top of it.

gchanan commented:
On your question, I don't think we should add calls to get random numbers directly to the generator; the normal pattern is to pass in the generator to a function that generates the random numbers, right?

rachtsingh commented:
Yeah, that sounds right. I'm not sure about casting the result of `unsafeGetTH` from `void *` to be able to sample though - is it ok to change the generator to output the specialized `Generator` when parsing `native_functions.yaml`?

rachtsingh commented on aten/src/ATen/cuda/CUDAApplyUtils.cuh:
```diff
@@ -283,12 +283,14 @@ bool CUDA_tensor_apply2(at::Tensor a,
+  //g if (!getApplyGrid(totalElements, grid)) {
+  //g   return false;
+  //g }
+  //g grid = dim3(1);
```
This is temporary - looking for advice on how to best limit the blockDim to 256 when calling `CUDA_tensor_apply2 `

rachtsingh commented:
Ok, I think it's probably ready for review. @fritzo, given the C++/CUDA changes, it's probably best to  ask someone from PyTorch to review as well? I don't know what exactly the plans are for CUDAGenerator, etc.

fritzo commented on torch/distributions/poisson.py:
```diff
@@ -0,0 +1,53 @@
+
+class Poisson(Distribution):
+    r"""
+    Creates a Poisson distribution parameterized by `_lambda`, the rate parameter.
```
nit: preceding underscore usually denotes a private variable. It's probably safer to call this `lam` or `lambda_`. Alternatively it would be nice to call it `rate` which is both semantically meaningful and improves compatibility with tensorflow distributions.

fritzo replied:
For example `self._lambda` would be the obvious member name but that looks private. Instead `self.lambda_` or `self.rate` are clearly public.

rachtsingh replied:
Both sound good to me. Let's use `rate` since that's more evocative? I.e. like `mean`/`std` in `Normal` already, and matches up with `Exponential`, which is the right idea.

fritzo commented:
@apaszke How should we test the CUDA sampler? Do we use the same tests in `test_distributions.py` that we use for CPU samplers, or should these new tests live in `test_cuda.py`?

apaszke commented:
I think `test_distributions.py` is ok. Just make `unittest` skip the tests if CUDA is unavailable (look at `test_nn.py` and `test_autograd.py` for examples).

apaszke commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -0,0 +1,74 @@
+} // at::native::dist
+
+Tensor _s_poisson_cuda(const Tensor& lambda, Generator* gen) {
+  Tensor ret = lambda.type().toScalarType(kDouble).zeros(lambda.sizes());
```
It's better to use `tensor` instead of `zeros` if you don't depend on the initial values of the tensor (it will be uninitialized)

apaszke commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -0,0 +1,74 @@
+
+Tensor _s_poisson_cuda(const Tensor& lambda, Generator* gen) {
+  Tensor ret = lambda.type().toScalarType(kDouble).zeros(lambda.sizes());
+  auto lambda_ = lambda.toType(ScalarType::Double);
```
Is this unstable in fp32? fp64 math is extremely slow on non-Tesla GPUs, and we should avoid it

apaszke commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -0,0 +1,74 @@
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
```
Is this the right place for this copyright notice?

apaszke commented on torch/csrc/Module.cpp:
```diff
@@ -750,6 +751,7 @@ static PyMethodDef TorchMethods[] = {
   {"_standard_gamma",  (PyCFunction)THPModule_standard_gamma,    METH_VARARGS | METH_KEYWORDS, NULL},
   {"_dirichlet_grad", (PyCFunction)THPModule_dirichlet_grad,    METH_VARARGS | METH_KEYWORDS, NULL},
   {"bernoulli",       (PyCFunction)THPModule_bernoulli,         METH_VARARGS | METH_KEYWORDS, NULL},
+  {"poisson",         (PyCFunction)THPModule_poisson,           METH_VARARGS | METH_KEYWORDS, NULL},
```
Let's avoid adding more of the sampling methods to the global namespace. `torch.distributions` is the way to expose those samplers

rachtsingh commented:
Ok, think I addressed comments (thanks for the review!) except for Adam's comment about cluttering the global namespace. What's the right way to do that? I think everything from `._C` gets imported automatically, so I'm not sure which config file to edit to expose it to Python.

gchanan commented on aten/src/ATen/native/cuda/Distributions.cu:
```diff
@@ -40,12 +31,13 @@ namespace dist {
   Tensor ret = lambda.type().tensor(lambda.sizes());
   auto lambda_ = lambda.toType(ScalarType::Float);
-  dispatch_all<void, dist::PoissonOpCUDA>(ret.type(), "poisson", ret, lambda_, dist::get_states(gen));
+  dispatch_floating_types<void, dist::PoissonOpCUDA>(ret.type(), "poisson", ret, lambda_, dist::get_states(gen));
```
if you are having trouble with half, try `DISPATCH_ALL_FLOATING_TYPES`; `dispatch_all` uses `Half` and `DISPATCH_ALL_FLOATING_TYPES` uses `half`; we should resolve this, probably always using `Half` (and providing __device__ conversion functions?).  CC @colesbury

rachtsingh replied:
Thanks, that works!

apaszke replied:
Ideally we would always use `half` for device functions and `at::Half` for host functions.

rachtsingh commented:
After doing some checks, it looks like `curand_poisson` without using the more robust API uses either an inverse CDF sampler or the transformed rejection method, neither of which are accurate for low values of `lambda`. So, it fails the distribution tests (good thing they work!). 

I think there's 3 options - (1) to merge as-is (by raising the threshold on the CUDA version), but to warn that for low values of `lambda`, the results may be biased, or (2) merge the CPU version, not the GPU version, or (3) punt on this until we get Philox-based PRNGs up, which was what I was planning to tackle next.

apaszke commented:
I'm fine with all options. I'll leave the decision to @fritzo

fritzo commented:
Option 1 sounds very reasonable. You could also [file an issue at probtorch/pytorch](https://github.com/probtorch/pytorch/issues/new) noting that we should revisit this.

rachtsingh commented:
OK, fixed the issues and added a Philox-based RNG. I don't think the sampling algorithm is the most accurate, but we can work on it in the future I think.

Also, I think there's a bug in the handling of `failure_rate` (note that I *decreased* that to make the tests pass), but I'll have to double check the math before fixing.

rachtsingh commented:
(discussing in https://github.com/pytorch/pytorch/pull/4556)
